src/trainer.py

```diff
@@ -0,0 +1,125 @@
+# __author__ = "Vasudev Gupta"
+# __author_email__ = "7vasudevgupta@gmail.com"
+
+"""Simple Trainer for handling TensorFlow Model Training"""
+
+from dataclasses import asdict, dataclass
+
+import tensorflow as tf
+import wandb
+from tqdm.auto import tqdm
+
+@dataclass
+class Trainer:
+    lr: float
+    max_epochs: int = 2
+
+    logging_steps: int = 10
+    eval_steps: int = 10
+
+    use_tpu: bool = False
+
+    save_steps: int = 10
+    ckpt_dir: str = "checkpoints" # None
+
+    # wandb args
+    project_name: str = "gsoc-wav2vec2"
+    run_name: str = "finetuning"
+
+    def setup(self, model, optimizer, restore_ckpt=False):
+
+        if self.use_tpu:
+            resolver = tf.distribute.cluster_resolver.TPUClusterResolver(tpu='')
+            tf.config.experimental_connect_to_cluster(resolver)
+            tf.tpu.experimental.initialize_tpu_system(resolver)
+            self.strategy = tf.distribute.TPUStrategy(resolver)
+            with self.strategy.scope():
+                self.model = model
+                self.optimizer = optimizer
+        else:
+            self.strategy = None
+            self.model = model
+            self.optimizer = optimizer
+
+        self.logger = wandb.init(
+            project_name=self.project_name, run_name=self.run_name, config=asdict(self)
+        )
+
+        checkpoint = tf.train.Checkpoint(model=self.model, optimizer=self.optimizer)
+        self.manager = tf.train.CheckpointManager(checkpoint, self.ckpt_dir, max_to_keep=5)
+
+        if restore_ckpt:
+            self.restore_checkpoint(checkpoint)
+
+    def train(self, tr_dataset_fn, val_dataset_fn):
```

Comment:
Prefer subclassing `tf.model.Model` and put these utilities there. It has many benefits. [Here's](https://keras.io/guides/customizing_what_happens_in_fit/) an overview.

Reply:
@sayakpaul is it very important to use `model.fit(...)`?? I am unable to understand the benefits of doing that (other than reducing some boilerplate).

I actually like to write raw trainer (not sure why 😅😅).

Reply:
Here are a few examples why we encourage using that:

* https://keras.io/examples/nlp/nl_image_search/
* https://www.kaggle.com/amyjang/monet-cyclegan-tutorial
* https://keras.io/examples/generative/cyclegan
* https://keras.io/examples/generative/wgan_gp/
* https://keras.io/examples/vision/simsiam/
* https://keras.io/examples/vision/knowledge_distillation/

The advantages of coding your training logic this way are not just about reducing boilerplate. If I wanted to plugin any fancy `tf.keras.callbacks` I can just specify those under my `callbacks` argument of `model.fit()`. If I wanted to make use of mixed-precision or distribution strategies, I can simply make use of `strategy.scope()` without having to worry about any sharding, fancy batch splitting, and so on. 

I will suggest studying the examples above first, get a feel of why we encourage implementing `model.fit()` and then ask anything about it.

Reply:
Okay. Thanks a lot for sharing resources and for a very quick reply. I will go through it.